### PR TITLE
FieldReference: add formal grammar definition docs

### PR DESCRIPTION
**Pull request by yaauie**
These are the docs extracted from #9543 and define the formal grammar that is implemented by strict-mode FieldReference parsing. It should be back-ported to 6.4 and 6.x.

## Background:

As initially implemented, the field reference syntax was only loosely defined in docs, and was coupled with a parser that silently did very unexpected things when encountering syntax that wasn't in the spirit of the documentation. To address this, we did several things:
 - in `6.3.x`, fix a bug on a specific set of bad inputs that could cause Logstash to crash; in these cases we reject input.
 - in `6.4`, add an opt-in strict mode that rejects invalid input; when using the normal *compat*-mode parser, _warn_ the user when input wouldn't pass the strict-mode but use the result from the old lenient parser.
- in `7.x`, the parser is _always_ strict and _only_ accepts input that is legal in the formal grammar.

The formal grammar defined in this PR is valid for _all_ branches, whether or not Logstash is configured to operate in *strict*-mode. The difference is that *compat*- and *legacy*-modes will _accept_ additional input that is outside of the formal grammar (and in these cases, the behaviour of inputs that do _not_ pass this formal grammar is undefined).

### CODE REVIEW

File: docs/static/field-reference.asciidoc [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  ;
+
+[[formal-grammar-event-api-composite-field-reference]]
+==== Composite Field Reference
+
+In some cases, it may be necessary to programmatically _compose_ a Field Reference from one or more Field References,
+such as when manipulating fields in a plugin or while using the Ruby Filter plugin and the Event API.
+
+[source,ruby,indent=4]
+    fieldReference = "[path][to][deep nested field]"
```
Comment (karenzone) [inline docs/static/field-reference.asciidoc]: WRT `indent=4` .   It appears that asciidoc does not support this notation.
When I get a successful build, we can see what it looks like and find a way to achieve the result you want using a different approach.

File: docs/static/field-reference.asciidoc [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+// `[foo][[bar]][bingo]`
+|`[foo]\[[bar]][bingo]`
+|`[foo][bar][bingo]`
+
+// `[[ok]]`
+|`\[[ok]]`
+|`[ok]`
+|===
+
```
Comment (karenzone) [inline docs/static/field-reference.asciidoc]: It appears that asciidoc doesn't like something about the table.  Here's the error: 
`asciidoc: WARNING: field-reference.asciidoc: line 109: nested inline passthrough`

I'll keep experimenting to see if I can make it work.  Ideas are welcome.

Comment (yaauie) [reply docs/static/field-reference.asciidoc]: Argh. Okay. I'll just add a second table in comments that is wholly separate.

Comment (yaauie) [reply docs/static/field-reference.asciidoc]: I think using the asciidoc "plus for passthrough" syntax to escape the entire strings instead of randomly injecting backslashes reads clearly enough that we don't need the individual comments.

**Comment by yaauie** [comment]
@karenzone I resolved all of the bits you commented on; can you take another look? :)

File: docs/static/configuration.asciidoc [hunk 3]
```diff
@@ -319,9 +319,9 @@ will not work in an input block.
 ==== Field References
 
 It is often useful to be able to refer to a field by name. To do this,
-you can use the Logstash field reference syntax.
+you can use the Logstash <<field-reference,field reference syntax>>.
```
[the remark this answers is not in the document]
Comment (karenzone) [reply docs/static/configuration.asciidoc]: I tend to think of it as the anchor ID for the section, but yes.  
screen shot 2018-07-26 at 5 28 23 pm

`[[field-references]]` is the anchor  and `<<field-references>>` is the corresponding link.

File: docs/static/field-reference.asciidoc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      # |                      |    |  |               |    |               | |
+      # +----string-argument---+    |  +--field-ref----+    +--field-ref----+ |
+      #                             +-------- sprintf format string ----------+
+      }
+    }
+  }
+}
+
+[[formal-grammar]]
+=== Formal Grammar
```
Comment (dedemorton) [inline docs/static/field-reference.asciidoc]: Who is the audience for this content? TBH, I find the details very hard to follow, maybe because I'm not familiar with the notation you're using here (antlr?) to represent the grammar. I'm concerned that users who aren't familiar with this style of notation will be confused by this topic. I don't think we can expect all of our users to have a programming background. If the goal is to help users understand how to use field references, I'd suggest providing more examples (maybe that show example event data and the syntax you would use to access specific fields). 

If a subset of users must know the formal grammar, then I'd consider adding it as an appendix rather than within the main flow of content. The way it's structured currently, the new topic breaks the book structure and doesn't fit logically into the hierarchy of topics.

Comment (yaauie) [reply docs/static/field-reference.asciidoc]: I agree that an appendix would be a useful place for this, the goal being that when someone runs into an issue where the loosey-goosey description of field references in other docs leads them to create inputs that fail to parse, they have somewhere to go to disambiguate while they debug.

How should I change this PR to make this an appendix entry?

***

I agree that the ANTLR-style grammar (or any formal meta-grammar) is a bit obtuse to an observer who is not familiar with formal grammar definitions, which is why each block is accompanied by an explanation in plain english using _Capital Emphasis_ to refer to proper nouns that are elsewhere defined in the document.

***

There may be value in re-arranging the sections to subtly get a reader into the habit of looking for more information in separate sections. 

 - The first section, _Field Reference Literal_, talks about _Path Fragments_, so it would be helpful if the _Path Fragment_ section immediately followed it.
 - The _Path Fragment_ section talks about _Field Names_, so it would be helpful if the _Field Name_ section immediately followed it.
  - A reader who gets this far will have much more success encountering the _Field Reference (Event APIs)_ section, which refers to _Field Reference Literal_  and  _Field Name_ (which they already understand), and _Composite Field Reference_ (which they don't know yet, and don't often _need_ to know, but should by now be familiar enough to guess that it is defined in this document should they need to drill in deeper).

Comment (dedemorton) [reply docs/static/field-reference.asciidoc]: @yaauie To format this as an appendix, you need to move the topic to the end of the book  and tag the section correctly. Should look something like this:

```
["appendix",id="your-section-id"]
== Title

Your content
...

```

**Comment by karenzone** [comment]
@yaauie   I included the goodness from this PR in #10057 .  Thanks for the good work here!
